tests/raftstore/pd.rs

```diff
@@ -82,114 +99,357 @@ impl Cluster {
-        }
 
-        for peer in region.get_peers() {
-            let store = self.stores.get_mut(&peer.get_store_id()).unwrap();
-            store.region_ids.insert(region.get_id());
+        let cur_region = self.get_region_by_id(region.get_id()).unwrap();
+
+        let cur_conf_ver = cur_region.get_region_epoch().get_conf_ver();
+        try!(check_stale_region(&cur_region, &region));
+
+        let region_peer_len = region.get_peers().len();
+        let cur_region_peer_len = cur_region.get_peers().len();
+
+        if conf_ver > cur_conf_ver {
+            // If ConfVer changed, TiKV has added/removed one peer already.
+            // So pd and TiKV can't have same peer number and can only have
+            // only one different peer.
+            // E.g, we can't meet following cases:
+            // 1) pd is (1, 2, 3), TiKV is (1)
+            // 2) pd is (1), TiKV is (1, 2, 3)
+            // 3) pd is (1, 2), TiKV is (3)
+            // 4) pd id (1), TiKV is (2, 3)
+
+            assert!(region_peer_len != cur_region_peer_len);
+
+            if cur_region_peer_len > region_peer_len {
+                // must pd is (1, 2), TiKV is (1)
+                assert_eq!(cur_region_peer_len - region_peer_len, 1);
+                let peers = setdiff_peers(&cur_region, &region);
+                assert_eq!(peers.len(), 1);
+                assert!(setdiff_peers(&region, &cur_region).is_empty());
+            } else {
+                // must pd is (1), TiKV is (1, 2)
+                assert_eq!(region_peer_len - cur_region_peer_len, 1);
+                let peers = setdiff_peers(&region, &cur_region);
+                assert_eq!(peers.len(), 1);
+                assert!(setdiff_peers(&cur_region, &region).is_empty());
+            }
+
+            // update the region.
+            assert!(self.regions.insert(end_key, region.clone()).is_some());
+        } else {
+            must_same_peers(&cur_region, &region);
         }
 
-        assert!(self.regions.insert(end_key, region).is_some());
+        let mut resp = pdpb::RegionHeartbeatResponse::new();
 
-        Ok(())
-    }
+        if self.rule.is_some() {
+            let rule = self.rule.as_ref().unwrap();
+            if let Some(change_peer) = rule(&region) {
+                resp.set_change_peer(change_peer);
+            }
 
-    fn split_region(&mut self, left: metapb::Region, right: metapb::Region) -> Result<()> {
-        let left_end_key = enc_end_key(&left);
-        let right_end_key = enc_end_key(&right);
+            return Ok(resp);
+        }
 
-        assert!(left.get_start_key() < left.get_end_key());
-        assert_eq!(left.get_end_key(), right.get_start_key());
-        assert!(right.get_end_key().is_empty() || right.get_start_key() < right.get_end_key());
+        // If no rule, or rule returns None, use default max_peer_number check.
+        let mut change_peer = pdpb::ChangePeer::new();
+
+        let max_peer_number = self.meta.get_max_peer_number() as usize;
```

Comment:
em, this will change kvproto too, I will fire another issue to change kvproto, and here we can rename var first.